examples/coap-server.c

```diff
@@ -1112,6 +1145,74 @@ static int cmdline_read_identity_check(char *arg) {
   return valid_ids.count > 0;
 }
 
+static int cmdline_read_pki_sni_check(char *arg) {
+  FILE *fp = fopen(arg, "r");
+  static char tmpbuf[256];
+  if (fp == NULL) {
+    coap_log(LOG_ERR, "SNI file: %s: Unable to open\n", arg);
+    return 0;
+  }
+  while (fgets(tmpbuf, sizeof(tmpbuf), fp) != NULL) {
+    char *cp = tmpbuf;
+    char *tcp = strchr(cp, '\n');
+
+    if (tmpbuf[0] == '#')
+      continue;
+    if (tcp)
+      *tcp = '\000';
+
+    tcp = strchr(cp, ',');
+    if (tcp) {
+      pki_sni_def_t *new_pki_sni_list;
+      new_pki_sni_list = realloc(valid_pki_snis.pki_sni_list,
+            (valid_pki_snis.count + 1)*sizeof (valid_pki_snis.pki_sni_list[0]));
+      if (new_pki_sni_list == NULL) {
+        break;
+      }
+      valid_pki_snis.pki_sni_list = new_pki_sni_list;
+      valid_pki_snis.pki_sni_list[valid_pki_snis.count].sni_match =
+                                                           strndup(cp, tcp-cp);
+      cp = tcp+1;
+      tcp = strchr(cp, ',');
+      if (tcp) {
+        struct stat statbuf;
+        int fail = 0;
+        valid_pki_snis.pki_sni_list[valid_pki_snis.count].new_cert =
+                             strndup(cp, tcp-cp);
+        cp = tcp+1;
+        valid_pki_snis.pki_sni_list[valid_pki_snis.count].new_ca =
+                             strndup(cp, strlen(cp));
+        if (stat(valid_pki_snis.pki_sni_list[valid_pki_snis.count].new_cert,
```

Comment:
This and the following `stat()` (in line 1191) are a bit fragile. We probably cannot avoid the TOCTOU here but we could at least check that we are looking at a regular file with read permissions. The easiest way to achieve this would be:

```c
   if (access(valid_pki_snis.pki_sni_list[valid_pki_snis.count].new_cert, R_OK) { .... }
   if (access(valid_pki_snis.pki_sni_list[valid_pki_snis.count].new_ca, R_OK) { .... }
```

Reply:
Updated using access()